File: longest_substring_without_repeating_characters/longest_substring_no_rep_char.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        max = 0
        dict = {}
        
        for i, s in enumerate(s):
            if s in dict:
                dict = {key:val for key, val in dict.items() if val >= dict[s]}

            dict[s] = i;
            max = len(dict) if max < len(dict) else max
        
                
        return max

File: longest_substring_without_repeating_characters/longest_substring_no_rep_char.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int res = 0, left = 0, i = 0, n = s.size();
        unordered_set<char> t;
        while (i < n) {
            if (!t.count(s[i])) {
                t.insert(s[i++]);
                res = max(res, (int)t.size());
            }  else {
                t.erase(s[left++]);
            }
        }
        return res;
    }
};
